core/state/statedb.go

```diff
@@ -728,6 +728,31 @@ func (s *StateDB) Copy() *StateDB {
 	if s.prefetcher != nil {
 		state.prefetcher = s.prefetcher.copy()
 	}
+	if s.snaps != nil {
+		// In order for the miner to be able to use and make additions
+		// to the snapshot tree, we need to copy that aswell.
+		// Otherwise, any block mined by ourselves will cause gaps in the tree,
+		// and force the miner to operate trie-backed only
+		state.snaps = s.snaps
+		state.snap = s.snap
+		// TODO: Should we copy or make new (empty) ones here?
+		// If we copy, we need to ensure concurrency safety.
+		// If we don't copy, we run the risk of consensus breaking.
+		// In theory, as the state is copied, it's still 'fresh', and these
+		// should be empty.
```

Comment:
The miner copies the state after running the transactions, but before claiming the block reward. It does this so it ca pile more txs on top. In that case, the state is fresh only if no tx populated it (i.e. post byzantium)